hazelcast/src/main/java/com/hazelcast/wan/merkletree/ArrayMerkleTree.java

```diff
@@ -238,11 +228,31 @@ private void updateBranch(int leafOrder) {
     private void addKeyToLeaf(int leafOrder, int keyHash, Object key) {
         int relativeLeafOrder = leafOrder - leafLevelOrder;
         leafKeys[relativeLeafOrder].add(key, keyHash);
+        updateFootprint();
     }
 
     private void removeKeyFromLeaf(int leafOrder, int keyHash, Object key) {
         int relativeLeafOrder = leafOrder - leafLevelOrder;
         leafKeys[relativeLeafOrder].remove(key, keyHash);
+        updateFootprint();
+    }
+
+    @SuppressWarnings("checkstyle:trailingcomment")
+    private void updateFootprint() {
```

Comment:
What is the reason in calculating the footprint on every addition/removal and not in the footprint itself, threading issues I guess? Looks like that may be pretty expensive if there is a lot of leaves.

Reply:
Yes, the Merkle trees are updated on the partition threads, while footprint is called from different threads. 

I'm thinking on how to make the update cheaper, like dealing only with the leaf being touched, so that no iteration is needed.